flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/CheckpointBarrierUnaligner.java

```diff
@@ -0,0 +1,338 @@
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.concurrent.CompletableFuture;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * {@link CheckpointBarrierUnaligner} is used for triggering checkpoint while reading the first barrier
+ * and keeping track of the number of received barriers and consumed barriers.
+ */
+@Internal
+public class CheckpointBarrierUnaligner extends CheckpointBarrierHandler {
+
+	private static final Logger LOG = LoggerFactory.getLogger(CheckpointBarrierUnaligner.class);
+	public static final CompletableFuture<Void> DONE = CompletableFuture.completedFuture(null);
+
+	private final String taskName;
+
+	/**
+	 * Tag the state of which input channel has read the barrier. If one channel has read the barrier by task,
+	 * the respective in-flight input buffers should be empty when triggering unaligned checkpoint.
+	 */
+	private final boolean[] barrierConsumedChannels;
+
+	/**
+	 * Tag the state of which input channel has received the barrier. Until the barrier of a specific channel is
+	 * received, all new buffers need to be persisted after unaligned checkpointing has been started.
+	 */
+	private final boolean[] barrierReceivedChannels;
+
+	/**
+	 * Contains the offsets of the channel indices for each gate when flattening the channels of all gates.
+	 *
+	 * <p>For example, consider 3 gates with 4 channels, {@code gateChannelOffsets = [0, 4, 8]}.
+	 */
+	private final int[] gateChannelOffsets;
+
+	/**
+	 * The number of input channels which has read the barrier by task.
+	 */
+	private int numBarriersReceived;
+
+	/**
+	 * The checkpoint id to guarantee that we would trigger only one checkpoint when reading the same barrier from
+	 * different channels.
+	 */
+	private long currentCheckpointId = -1L;
+
+	/** The number of opened channels. */
+	private int numOpenChannels;
+
+	/** A future indicating that all barriers of the a given checkpoint have been read. */
+	private CompletableFuture<Void> allBarriersReceivedFuture = DONE;
+
+	private final ChannelStateWriter channelStateWriter;
+
+	CheckpointBarrierUnaligner(
+			int[] numberOfInputChannelsPerGate,
+			ChannelStateWriter channelStateWriter,
+			String taskName,
+			AbstractInvokable toNotifyOnCheckpoint) {
+		super(toNotifyOnCheckpoint);
+
+		this.taskName = taskName;
+		this.channelStateWriter = checkNotNull(channelStateWriter);
+
+		final int numGates = numberOfInputChannelsPerGate.length;
+
+		gateChannelOffsets = new int[numGates];
+		for (int index = 1; index < numGates; index++) {
+			gateChannelOffsets[index] = gateChannelOffsets[index - 1] + numberOfInputChannelsPerGate[index - 1];
+		}
+
+		final int totalNumChannels = gateChannelOffsets[numGates - 1] + numberOfInputChannelsPerGate[numGates - 1];
+		barrierConsumedChannels = new boolean[totalNumChannels];
+		Arrays.fill(barrierConsumedChannels, true);
```

Comment:
The default value should be false for `barrierConsumedChannels`, otherwise it seems inconsistent with the variable semantic. Only after task processing the barrier from respective channel, it should be set true.

Reply:
Since current checkpoint id is -1. `true` is closer to the semantics as we don't ever expect a(nother) barrier with that value to be consumed.
`CheckpointedInputGate#getInFlightData` returns an empty list as expected iff we set it to `true`.